Source/DafnyLanguageServer/Workspace/DocumentDatabase.cs

```diff
@@ -114,65 +116,111 @@ ISymbolTableRelocator symbolTableRelocator
 
       databaseEntry.CancelPendingUpdates();
       var cancellationSource = new CancellationTokenSource();
+      var previousDocumentTask = FirstSuccessfulAsync(databaseEntry.VerifiedDocument, databaseEntry.ResolvedDocument);
+      var resolvedDocumentTask = ApplyChangesAsync(previousDocumentTask, documentChange, cancellationSource.Token);
+      var verifiedDocument = VerifyAsync(resolvedDocumentTask, VerifyOnChange, cancellationSource.Token);
       var updatedEntry = new DocumentEntry(
         documentChange.TextDocument.Version,
-        Task.Run(async () => await ApplyChangesAsync(await databaseEntry.Document, documentChange, cancellationSource.Token)),
+        resolvedDocumentTask,
+        verifiedDocument,
         cancellationSource
       );
       documents[documentUri] = updatedEntry;
-      return await updatedEntry.Document;
+      return resolvedDocumentTask.ToObservable().Concat(verifiedDocument.ToObservable());
+    }
+
+
+    /// <summary>
+    /// Returns the first task that completes successfully.
+    /// If multiple tasks are already completed, the first one is returned.
+    /// </summary>
+    private static Task<T> FirstSuccessfulAsync<T>(params Task<T>[] tasks) {
+      var taskList = tasks.ToList();
+      var tcs = new TaskCompletionSource<T>();
+      int remainingTasks = taskList.Count;
+      foreach (var task in taskList) {
+        if (tcs.Task.IsCompleted) {
+          break;
+        }
+
+        _ = task.ContinueWith(t => {
+          if (task.Status == TaskStatus.RanToCompletion) {
+            tcs.TrySetResult(t.Result);
+          } else if (Interlocked.Decrement(ref remainingTasks) == 0) {
```

Comment:
Wouldn't the checks of the other tasks be unnecessary if you handle the failure of the current task here?